File: performance_validation/competitors/cpp/json_parser.cpp
// C++ JSON parser benchmark implementation for fair comparison with Seen
// High-performance JSON parser using similar architecture to Seen implementation

#include <iostream>
#include <string>
#include <vector>
#include <unordered_map>
#include <chrono>
#include <fstream>
#include <filesystem>
#include <sstream>
#include <memory>
#include <variant>
#include <iomanip>
#include <cctype>
#include <stdexcept>

// Forward declaration
class JsonValue;

using JsonObject = std::unordered_map<std::string, JsonValue>;
using JsonArray = std::vector<JsonValue>;

class JsonValue {
public:
    using Value = std::variant<
        std::nullptr_t,
        bool,
        double,
        std::string,
        JsonArray,
        JsonObject
    >;

    Value value;

    JsonValue() : value(nullptr) {}
    JsonValue(std::nullptr_t) : value(nullptr) {}
    JsonValue(bool b) : value(b) {}
    JsonValue(double d) : value(d) {}
    JsonValue(int i) : value(static_cast<double>(i)) {}
    JsonValue(const std::string& s) : value(s) {}
    JsonValue(std::string&& s) : value(std::move(s)) {}
    JsonValue(const JsonArray& arr) : value(arr) {}
    JsonValue(JsonArray&& arr) : value(std::move(arr)) {}
    JsonValue(const JsonObject& obj) : value(obj) {}
    JsonValue(JsonObject&& obj) : value(std::move(obj)) {}

    bool isNull() const { return std::holds_alternative<std::nullptr_t>(value); }
    bool isBool() const { return std::holds_alternative<bool>(value); }
    bool isNumber() const { return std::holds_alternative<double>(value); }
    bool isString() const { return std::holds_alternative<std::string>(value); }
    bool isArray() const { return std::holds_alternative<JsonArray>(value); }
    bool isObject() const { return std::holds_alternative<JsonObject>(value); }

    bool asBool() const { return std::get<bool>(value); }
    double asNumber() const { return std::get<double>(value); }
    const std::string& asString() const { return std::get<std::string>(value); }
    const JsonArray& asArray() const { return std::get<JsonArray>(value); }
    const JsonObject& asObject() const { return std::get<JsonObject>(value); }

    bool isValid() const {
        return std::visit([](const auto& v) -> bool {
            using T = std::decay_t<decltype(v)>;
            if constexpr (std::is_same_v<T, std::nullptr_t>) {
                return true;
            } else if constexpr (std::is_same_v<T, bool>) {
                return true;
            } else if constexpr (std::is_same_v<T, double>) {
                return std::isfinite(v);
            } else if constexpr (std::is_same_v<T, std::string>) {
                return !v.empty();
            } else if constexpr (std::is_same_v<T, JsonArray>) {
                return std::all_of(v.begin(), v.end(), [](const JsonValue& item) {
                    return item.isValid();
                });
            } else if constexpr (std::is_same_v<T, JsonObject>) {
                return std::all_of(v.begin(), v.end(), [](const auto& pair) {
                    return pair.second.isValid();
                });
            }
            return false;
        }, value);
    }

    size_t size() const {
        return std::visit([](const auto& v) -> size_t {
            using T = std::decay_t<decltype(v)>;
            if constexpr (std::is_same_v<T, std::nullptr_t> || 
                         std::is_same_v<T, bool> || 
                         std::is_same_v<T, double>) {
                return 1;
            } else if constexpr (std::is_same_v<T, std::string>) {
                return v.length();
            } else if constexpr (std::is_same_v<T, JsonArray>) {
                size_t total = 0;
                for (const auto& item : v) {
                    total += item.size();
                }
                return total;
            } else if constexpr (std::is_same_v<T, JsonObject>) {
                size_t total = v.size(); // Count keys
                for (const auto& pair : v) {
                    total += pair.second.size();
                }
                return total;
            }
            return 0;
        }, value);
    }
};

class JsonParser {
private:
    std::string input;
    size_t position;
    size_t line;
    size_t column;

    bool isAtEnd() const {
        return position >= input.length();
    }

    char currentChar() const {
        if (isAtEnd()) return '\0';
        return input[position];
    }

    char peekChar() const {
        if (position + 1 >= input.length()) return '\0';
        return input[position + 1];
    }

    char advance() {
        if (isAtEnd()) return '\0';
        
        char ch = input[position++];
        
        if (ch == '\n') {
            line++;
            column = 1;
        } else {
            column++;
        }
        
        return ch;
    }

    void skipWhitespace() {
        while (!isAtEnd() && std::isspace(currentChar())) {
            advance();
        }
    }

    std::string parseUnicodeEscape() {
        std::string hex;
        
        for (int i = 0; i < 4; ++i) {
            if (isAtEnd() || !std::isxdigit(currentChar())) {
                throw std::runtime_error("Invalid unicode escape sequence at line " + 
                                       std::to_string(line) + ", column " + std::to_string(column));
            }
            hex += advance();
        }
        
        unsigned int codePoint = std::stoul(hex, nullptr, 16);
        
        if (codePoint <= 0x7F) {
            return std::string(1, static_cast<char>(codePoint));
        } else if (codePoint <= 0x7FF) {
            std::string result;
            result += static_cast<char>(0xC0 | (codePoint >> 6));
            result += static_cast<char>(0x80 | (codePoint & 0x3F));
            return result;
        } else if (codePoint <= 0xFFFF) {
            std::string result;
            result += static_cast<char>(0xE0 | (codePoint >> 12));
            result += static_cast<char>(0x80 | ((codePoint >> 6) & 0x3F));
            result += static_cast<char>(0x80 | (codePoint & 0x3F));
            return result;
        } else {
            // For simplicity, we'll use replacement character for higher code points
            return "\uFFFD";
        }
    }

    JsonValue parseString() {
        advance(); // consume opening quote
        
        std::string value;
        
        while (!isAtEnd() && currentChar() != '"') {
            char ch = currentChar();
            
            if (ch == '\\') {
                advance(); // consume backslash
                if (isAtEnd()) {
                    throw std::runtime_error("Unexpected end of input in string");
                }
                
                char escaped = currentChar();
                switch (escaped) {
                    case '"': value += '"'; break;
                    case '\\': value += '\\'; break;
                    case '/': value += '/'; break;
                    case 'b': value += '\b'; break;
                    case 'f': value += '\f'; break;
                    case 'n': value += '\n'; break;
                    case 'r': value += '\r'; break;
                    case 't': value += '\t'; break;
                    case 'u': {
                        advance(); // consume 'u'
                        value += parseUnicodeEscape();
                        continue; // Don't advance again
                    }
                    default:
                        throw std::runtime_error("Invalid escape sequence '\\" + 
                                               std::string(1, escaped) + "' at line " + 
                                               std::to_string(line) + ", column " + std::to_string(column));
                }
                advance();
            } else {
                value += ch;
                advance();
            }
        }
        
        if (isAtEnd() || currentChar() != '"') {
            throw std::runtime_error("Unterminated string");
        }
        
        advance(); // consume closing quote
        return JsonValue(std::move(value));
    }

    JsonValue parseNumber() {
        std::string number;
        
        // Handle negative sign
        if (currentChar() == '-') {
            number += advance();
        }
        
        // Parse integer part
        if (currentChar() == '0') {
            number += advance();
        } else if (std::isdigit(currentChar())) {
            while (!isAtEnd() && std::isdigit(currentChar())) {
                number += advance();
            }
        } else {
            throw std::runtime_error("Invalid number: missing digits");
        }
        
        // Parse decimal part
        if (!isAtEnd() && currentChar() == '.') {
            number += advance();
            
            if (isAtEnd() || !std::isdigit(currentChar())) {
                throw std::runtime_error("Invalid number: missing digits after decimal point");
            }
            
            while (!isAtEnd() && std::isdigit(currentChar())) {
                number += advance();
            }
        }
        
        // Parse exponent part
        if (!isAtEnd() && (currentChar() == 'e' || currentChar() == 'E')) {
            number += advance();
            
            if (!isAtEnd() && (currentChar() == '+' || currentChar() == '-')) {
                number += advance();
            }
            
            if (isAtEnd() || !std::isdigit(currentChar())) {
                throw std::runtime_error("Invalid number: missing digits in exponent");
            }
            
            while (!isAtEnd() && std::isdigit(currentChar())) {
                number += advance();
            }
        }
        
        try {
            double value = std::stod(number);
            return JsonValue(value);
        } catch (const std::exception&) {
            throw std::runtime_error("Invalid number format: '" + number + "'");
        }
    }

    bool matchKeyword(const std::string& keyword) {
        if (input.substr(position, keyword.length()) == keyword) {
            // Check that the keyword is not part of a larger identifier
            size_t nextPos = position + keyword.length();
            if (nextPos >= input.length() || !std::isalnum(input[nextPos])) {
                for (size_t i = 0; i < keyword.length(); ++i) {
                    advance();
                }
                return true;
            }
        }
        return false;
    }

    JsonValue parseBoolean() {
        if (matchKeyword("true")) {
            return JsonValue(true);
        } else if (matchKeyword("false")) {
            return JsonValue(false);
        } else {
            throw std::runtime_error("Invalid boolean at line " + 
                                   std::to_string(line) + ", column " + std::to_string(column));
        }
    }

    JsonValue parseNull() {
        if (matchKeyword("null")) {
            return JsonValue(nullptr);
        } else {
            throw std::runtime_error("Invalid null at line " + 
                                   std::to_string(line) + ", column " + std::to_string(column));
        }
    }

    JsonValue parseArray() {
        advance(); // consume '['
        skipWhitespace();
        
        JsonArray elements;
        
        // Handle empty array
        if (!isAtEnd() && currentChar() == ']') {
            advance();
            return JsonValue(std::move(elements));
        }
        
        while (true) {
            elements.push_back(parseValue());
            
            skipWhitespace();
            
            if (isAtEnd()) {
                throw std::runtime_error("Unexpected end of input in array");
            }
            
            char ch = currentChar();
            if (ch == ',') {
                advance();
                skipWhitespace();
            } else if (ch == ']') {
                advance();
                break;
            } else {
                throw std::runtime_error("Expected ',' or ']' but found '" + 
                                       std::string(1, ch) + "' at line " + 
                                       std::to_string(line) + ", column " + std::to_string(column));
            }
        }
        
        return JsonValue(std::move(elements));
    }

    JsonValue parseObject() {
        advance(); // consume '{'
        skipWhitespace();
        
        JsonObject object;
        
        // Handle empty object
        if (!isAtEnd() && currentChar() == '}') {
            advance();
            return JsonValue(std::move(object));
        }
        
        while (true) {
            // Parse key (must be string)
            JsonValue keyValue = parseString();
            if (!keyValue.isString()) {
                throw std::runtime_error("Object key must be a string");
            }
            std::string key = keyValue.asString();
            
            skipWhitespace();
            
            // Expect colon
            if (isAtEnd() || currentChar() != ':') {
                throw std::runtime_error("Expected ':' after object key at line " + 
                                       std::to_string(line) + ", column " + std::to_string(column));
            }
            advance();
            
            skipWhitespace();
            
            // Parse value
            JsonValue value = parseValue();
            object[std::move(key)] = std::move(value);
            
            skipWhitespace();
            
            if (isAtEnd()) {
                throw std::runtime_error("Unexpected end of input in object");
            }
            
            char ch = currentChar();
            if (ch == ',') {
                advance();
                skipWhitespace();
            } else if (ch == '}') {
                advance();
                break;
            } else {
                throw std::runtime_error("Expected ',' or '}' but found '" + 
                                       std::string(1, ch) + "' at line " + 
                                       std::to_string(line) + ", column " + std::to_string(column));
            }
        }
        
        return JsonValue(std::move(object));
    }

    JsonValue parseValue() {
        skipWhitespace();
        
        if (isAtEnd()) {
            throw std::runtime_error("Unexpected end of input");
        }
        
        char ch = currentChar();
        
        switch (ch) {
            case '"':
                return parseString();
            case '[':
                return parseArray();
            case '{':
                return parseObject();
            case 't':
            case 'f':
                return parseBoolean();
            case 'n':
                return parseNull();
            default:
                if (ch == '-' || std::isdigit(ch)) {
                    return parseNumber();
                } else {
                    throw std::runtime_error("Unexpected character '" + 
                                           std::string(1, ch) + "' at line " + 
                                           std::to_string(line) + ", column " + std::to_string(column));
                }
        }
    }

public:
    JsonParser(const std::string& input) 
        : input(input), position(0), line(1), column(1) {}

    JsonValue parse() {
        JsonValue value = parseValue();
        skipWhitespace();
        
        if (!isAtEnd()) {
            throw std::runtime_error("Unexpected content after JSON value at line " + 
                                   std::to_string(line) + ", column " + std::to_string(column));
        }
        
        return value;
    }
};

// Utility functions
std::string readFile(const std::string& filename) {
    std::ifstream file(filename);
    if (!file.is_open()) {
        throw std::runtime_error("Cannot open file: " + filename);
    }
    
    std::stringstream buffer;
    buffer << file.rdbuf();
    return buffer.str();
}

std::string generateDeeplyNestedJson(size_t depth) {
    std::string json;
    
    for (size_t i = 0; i < depth; ++i) {
        json += "{\"nested\":";
    }
    
    json += "\"value\"";
    
    for (size_t i = 0; i < depth; ++i) {
        json += "}";
    }
    
    return json;
}

std::string generateWideJson(size_t count) {
    std::string json = "{";
    
    for (size_t i = 0; i < count; ++i) {
        if (i > 0) {
            json += ",";
        }
        json += "\"key" + std::to_string(i) + "\":" + std::to_string(i);
    }
    
    json += "}";
    return json;
}

void benchmarkJsonParserRealWorld() {
    std::vector<std::string> testFiles = {
        "../../test_data/json_files/twitter.json",
        "../../test_data/json_files/canada.json",
        "../../test_data/json_files/citm_catalog.json",
        "../../test_data/json_files/large.json"
    };
    
    uint64_t totalElements = 0;
    uint64_t totalBytes = 0;
    std::chrono::duration<double> totalTime{0};
    
    for (const auto& filePath : testFiles) {
        if (!std::filesystem::exists(filePath)) {
            std::cout << "Warning: Test file " << filePath << " not found, skipping..." << std::endl;
            continue;
        }
        
        try {
            std::string content = readFile(filePath);
            size_t fileSize = content.size();
            
            std::cout << "Testing C++ JSON parser performance on " << filePath 
                      << " (" << fileSize << " bytes)" << std::endl;
            
            // Run multiple iterations for statistical accuracy
            const int iterations = 10;
            size_t fileElements = 0;
            std::chrono::duration<double> fileTime{0};
            
            for (int i = 0; i < iterations; ++i) {
                JsonParser parser(content);
                
                auto startTime = std::chrono::high_resolution_clock::now();
                JsonValue result = parser.parse();
                auto endTime = std::chrono::high_resolution_clock::now();
                
                if (!result.isValid()) {
                    throw std::runtime_error("Invalid JSON result");
                }
                
                fileElements = result.size();
                fileTime += endTime - startTime;
            }
            
            auto avgTime = fileTime / iterations;
            double bytesPerSecond = avgTime.count() > 0.0 ? 
                static_cast<double>(fileSize) / avgTime.count() : 0.0;
            
            std::cout << "  Elements: " << fileElements 
                      << ", Avg Time: " << std::fixed << std::setprecision(6) << avgTime.count() << "s"
                      << ", Bytes/sec: " << std::fixed << std::setprecision(0) << bytesPerSecond << std::endl;
            
            totalElements += fileElements;
            totalBytes += fileSize;
            totalTime += avgTime;
            
        } catch (const std::exception& e) {
            std::cerr << "Error processing " << filePath << ": " << e.what() << std::endl;
        }
    }
    
    double overallBytesPerSec = totalTime.count() > 0.0 ? 
        static_cast<double>(totalBytes) / totalTime.count() : 0.0;
    
    std::cout << "\nC++ JSON Parser Overall Performance:" << std::endl;
    std::cout << "  Total elements: " << totalElements << std::endl;
    std::cout << "  Total bytes: " << totalBytes << std::endl;
    std::cout << "  Total time: " << std::fixed << std::setprecision(6) << totalTime.count() << "s" << std::endl;
    std::cout << "  Average bytes/second: " << std::fixed << std::setprecision(0) << overallBytesPerSec << std::endl;
    std::cout << "  Average MB/sec: " << std::fixed << std::setprecision(2) << overallBytesPerSec / (1024.0 * 1024.0) << std::endl;
}

void benchmarkJsonParserStressTest() {
    std::cout << "Running C++ JSON parser stress tests..." << std::endl;
    
    try {
        // Test deeply nested structures
        std::string deeplyNested = generateDeeplyNestedJson(1000);
        JsonParser parser1(deeplyNested);
        
        auto start = std::chrono::high_resolution_clock::now();
        JsonValue result1 = parser1.parse();
        auto elapsed = std::chrono::high_resolution_clock::now() - start;
        
        if (!result1.isValid()) {
            throw std::runtime_error("Invalid deeply nested JSON result");
        }
        
        std::cout << "  Deeply nested (1000 levels): " 
                  << std::chrono::duration_cast<std::chrono::microseconds>(elapsed).count() 
                  << "μs" << std::endl;
        
        // Test wide structures
        std::string wideStructure = generateWideJson(10000);
        JsonParser parser2(wideStructure);
        
        start = std::chrono::high_resolution_clock::now();
        JsonValue result2 = parser2.parse();
        elapsed = std::chrono::high_resolution_clock::now() - start;
        
        if (!result2.isValid()) {
            throw std::runtime_error("Invalid wide JSON result");
        }
        
        std::cout << "  Wide structure (10000 keys): " 
                  << std::chrono::duration_cast<std::chrono::microseconds>(elapsed).count() 
                  << "μs" << std::endl;
        
    } catch (const std::exception& e) {
        std::cerr << "Error in stress test: " << e.what() << std::endl;
    }
}

int main() {
    std::cout << "Running C++ JSON Parser Benchmarks..." << std::endl;
    
    try {
        benchmarkJsonParserRealWorld();
        benchmarkJsonParserStressTest();
        
        std::cout << "C++ JSON parser benchmarks completed successfully!" << std::endl;
        return 0;
    } catch (const std::exception& e) {
        std::cerr << "Error running C++ JSON parser benchmark: " << e.what() << std::endl;
        return 1;
    }
}

File: performance_validation/competitors/rust/json_parser.rs
// Rust JSON parser benchmark implementation for fair comparison with Seen
// High-performance JSON parser using similar architecture to Seen implementation

use std::collections::HashMap;
use std::fs;
use std::time::{Duration, Instant};
use std::path::Path;
use std::str::Chars;
use std::iter::Peekable;

#[derive(Debug, Clone, PartialEq)]
pub enum JsonValue {
    Null,
    Bool(bool),
    Number(f64),
    String(String),
    Array(Vec<JsonValue>),
    Object(HashMap<String, JsonValue>),
}

impl JsonValue {
    pub fn is_valid(&self) -> bool {
        match self {
            JsonValue::Null => true,
            JsonValue::Bool(_) => true,
            JsonValue::Number(n) => n.is_finite(),
            JsonValue::String(s) => !s.is_empty(),
            JsonValue::Array(arr) => arr.iter().all(|v| v.is_valid()),
            JsonValue::Object(obj) => obj.values().all(|v| v.is_valid()),
        }
    }
    
    pub fn size(&self) -> usize {
        match self {
            JsonValue::Null => 1,
            JsonValue::Bool(_) => 1,
            JsonValue::Number(_) => 1,
            JsonValue::String(s) => s.len(),
            JsonValue::Array(arr) => arr.iter().map(|v| v.size()).sum(),
            JsonValue::Object(obj) => obj.values().map(|v| v.size()).sum::<usize>() + obj.len(),
        }
    }
}

pub struct JsonParser {
    chars: Peekable<Chars<'static>>,
    line: usize,
    column: usize,
    input: &'static str,
}

impl JsonParser {
    pub fn new(input: &'static str) -> Self {
        Self {
            chars: input.chars().peekable(),
            line: 1,
            column: 1,
            input,
        }
    }
    
    fn current_char(&mut self) -> Option<&char> {
        self.chars.peek()
    }
    
    fn advance(&mut self) -> Option<char> {
        match self.chars.next() {
            Some('\n') => {
                self.line += 1;
                self.column = 1;
                Some('\n')
            }
            Some(ch) => {
                self.column += 1;
                Some(ch)
            }
            None => None,
        }
    }
    
    fn skip_whitespace(&mut self) {
        while let Some(&ch) = self.current_char() {
            if ch.is_whitespace() {
                self.advance();
            } else {
                break;
            }
        }
    }
    
    pub fn parse(&mut self) -> Result<JsonValue, String> {
        let value = self.parse_value()?;
        self.skip_whitespace();
        
        if self.current_char().is_some() {
            return Err(format!("Unexpected content after JSON value at line {}, column {}", 
                              self.line, self.column));
        }
        
        Ok(value)
    }
    
    fn parse_value(&mut self) -> Result<JsonValue, String> {
        self.skip_whitespace();
        
        match self.current_char() {
            Some(&'"') => self.parse_string(),
            Some(&'[') => self.parse_array(),
            Some(&'{') => self.parse_object(),
            Some(&'t') | Some(&'f') => self.parse_boolean(),
            Some(&'n') => self.parse_null(),
            Some(&ch) if ch.is_ascii_digit() || ch == '-' => self.parse_number(),
            Some(&ch) => Err(format!("Unexpected character '{}' at line {}, column {}", 
                                   ch, self.line, self.column)),
            None => Err("Unexpected end of input".to_string()),
        }
    }
    
    fn parse_string(&mut self) -> Result<JsonValue, String> {
        self.advance(); // consume opening quote
        
        let mut value = String::new();
        
        while let Some(ch) = self.advance() {
            match ch {
                '"' => return Ok(JsonValue::String(value)),
                '\\' => {
                    match self.advance() {
                        Some('"') => value.push('"'),
                        Some('\\') => value.push('\\'),
                        Some('/') => value.push('/'),
                        Some('b') => value.push('\u{0008}'),
                        Some('f') => value.push('\u{000C}'),
                        Some('n') => value.push('\n'),
                        Some('r') => value.push('\r'),
                        Some('t') => value.push('\t'),
                        Some('u') => {
                            let hex = self.parse_unicode_escape()?;
                            value.push(hex);
                        }
                        Some(ch) => return Err(format!("Invalid escape sequence '\\{}' at line {}, column {}", 
                                                     ch, self.line, self.column)),
                        None => return Err("Unexpected end of input in string".to_string()),
                    }
                }
                ch => value.push(ch),
            }
        }
        
        Err("Unterminated string".to_string())
    }
    
    fn parse_unicode_escape(&mut self) -> Result<char, String> {
        let mut hex = String::new();
        
        for _ in 0..4 {
            match self.advance() {
                Some(ch) if ch.is_ascii_hexdigit() => hex.push(ch),
                Some(ch) => return Err(format!("Invalid hex digit '{}' in unicode escape at line {}, column {}", 
                                             ch, self.line, self.column)),
                None => return Err("Unexpected end of input in unicode escape".to_string()),
            }
        }
        
        let code_point = u32::from_str_radix(&hex, 16)
            .map_err(|_| "Invalid unicode escape sequence".to_string())?;
        
        char::from_u32(code_point)
            .ok_or("Invalid unicode code point".to_string())
    }
    
    fn parse_number(&mut self) -> Result<JsonValue, String> {
        let mut number = String::new();
        
        // Handle negative sign
        if let Some(&'-') = self.current_char() {
            number.push('-');
            self.advance();
        }
        
        // Parse integer part
        if let Some(&'0') = self.current_char() {
            number.push('0');
            self.advance();
        } else {
            let mut has_digits = false;
            while let Some(&ch) = self.current_char() {
                if ch.is_ascii_digit() {
                    number.push(ch);
                    self.advance();
                    has_digits = true;
                } else {
                    break;
                }
            }
            if !has_digits {
                return Err("Invalid number: missing digits".to_string());
            }
        }
        
        // Parse decimal part
        if let Some(&'.') = self.current_char() {
            number.push('.');
            self.advance();
            
            let mut has_digits = false;
            while let Some(&ch) = self.current_char() {
                if ch.is_ascii_digit() {
                    number.push(ch);
                    self.advance();
                    has_digits = true;
                } else {
                    break;
                }
            }
            
            if !has_digits {
                return Err("Invalid number: missing digits after decimal point".to_string());
            }
        }
        
        // Parse exponent part
        if let Some(&ch) = self.current_char() {
            if ch == 'e' || ch == 'E' {
                number.push('e');
                self.advance();
                
                if let Some(&sign) = self.current_char() {
                    if sign == '+' || sign == '-' {
                        number.push(sign);
                        self.advance();
                    }
                }
                
                let mut has_digits = false;
                while let Some(&ch) = self.current_char() {
                    if ch.is_ascii_digit() {
                        number.push(ch);
                        self.advance();
                        has_digits = true;
                    } else {
                        break;
                    }
                }
                
                if !has_digits {
                    return Err("Invalid number: missing digits in exponent".to_string());
                }
            }
        }
        
        let value: f64 = number.parse()
            .map_err(|_| format!("Invalid number format: '{}'", number))?;
        
        Ok(JsonValue::Number(value))
    }
    
    fn parse_boolean(&mut self) -> Result<JsonValue, String> {
        if self.match_keyword("true") {
            Ok(JsonValue::Bool(true))
        } else if self.match_keyword("false") {
            Ok(JsonValue::Bool(false))
        } else {
            Err(format!("Invalid boolean at line {}, column {}", self.line, self.column))
        }
    }
    
    fn parse_null(&mut self) -> Result<JsonValue, String> {
        if self.match_keyword("null") {
            Ok(JsonValue::Null)
        } else {
            Err(format!("Invalid null at line {}, column {}", self.line, self.column))
        }
    }
    
    fn match_keyword(&mut self, keyword: &str) -> bool {
        let mut temp_chars = self.chars.clone();
        
        for expected_char in keyword.chars() {
            match temp_chars.next() {
                Some(ch) if ch == expected_char => continue,
                _ => return false,
            }
        }
        
        // Check that the keyword is not part of a larger identifier
        if let Some(&ch) = temp_chars.peek() {
            if ch.is_alphanumeric() {
                return false;
            }
        }
        
        // Consume the keyword
        for _ in keyword.chars() {
            self.advance();
        }
        
        true
    }
    
    fn parse_array(&mut self) -> Result<JsonValue, String> {
        self.advance(); // consume '['
        self.skip_whitespace();
        
        let mut elements = Vec::new();
        
        // Handle empty array
        if let Some(&']') = self.current_char() {
            self.advance();
            return Ok(JsonValue::Array(elements));
        }
        
        loop {
            let value = self.parse_value()?;
            elements.push(value);
            
            self.skip_whitespace();
            
            match self.current_char() {
                Some(&',') => {
                    self.advance();
                    self.skip_whitespace();
                }
                Some(&']') => {
                    self.advance();
                    break;
                }
                Some(&ch) => return Err(format!("Expected ',' or ']' but found '{}' at line {}, column {}", 
                                              ch, self.line, self.column)),
                None => return Err("Unexpected end of input in array".to_string()),
            }
        }
        
        Ok(JsonValue::Array(elements))
    }
    
    fn parse_object(&mut self) -> Result<JsonValue, String> {
        self.advance(); // consume '{'
        self.skip_whitespace();
        
        let mut object = HashMap::new();
        
        // Handle empty object
        if let Some(&'}') = self.current_char() {
            self.advance();
            return Ok(JsonValue::Object(object));
        }
        
        loop {
            // Parse key (must be string)
            let key = match self.parse_string()? {
                JsonValue::String(s) => s,
                _ => return Err("Object key must be a string".to_string()),
            };
            
            self.skip_whitespace();
            
            // Expect colon
            match self.current_char() {
                Some(&':') => self.advance(),
                Some(&ch) => return Err(format!("Expected ':' after object key but found '{}' at line {}, column {}", 
                                              ch, self.line, self.column)),
                None => return Err("Unexpected end of input in object".to_string()),
            }
            
            self.skip_whitespace();
            
            // Parse value
            let value = self.parse_value()?;
            object.insert(key, value);
            
            self.skip_whitespace();
            
            match self.current_char() {
                Some(&',') => {
                    self.advance();
                    self.skip_whitespace();
                }
                Some(&'}') => {
                    self.advance();
                    break;
                }
                Some(&ch) => return Err(format!("Expected ',' or '}}' but found '{}' at line {}, column {}", 
                                              ch, self.line, self.column)),
                None => return Err("Unexpected end of input in object".to_string()),
            }
        }
        
        Ok(JsonValue::Object(object))
    }
}

// Benchmark functions
pub fn benchmark_json_parser_real_world() -> Result<(), Box<dyn std::error::Error>> {
    let test_files = vec![
        "../../test_data/json_files/twitter.json",
        "../../test_data/json_files/canada.json", 
        "../../test_data/json_files/citm_catalog.json",
        "../../test_data/json_files/large.json",
    ];
    
    let mut total_elements = 0u64;
    let mut total_bytes = 0u64;
    let mut total_time = Duration::new(0, 0);
    
    for file_path in test_files {
        if Path::new(file_path).exists() {
            let content = fs::read_to_string(file_path)?;
            let file_size = content.len();
            
            // We need to leak the string to get a 'static reference for the parser
            let static_content: &'static str = Box::leak(content.into_boxed_str());
            
            println!("Testing Rust JSON parser performance on {} ({} bytes)", file_path, file_size);
            
            // Run multiple iterations for statistical accuracy
            let iterations = 10;
            let mut file_elements = 0;
            let mut file_time = Duration::new(0, 0);
            
            for _ in 0..iterations {
                let mut parser = JsonParser::new(static_content);
                
                let start_time = Instant::now();
                let result = parser.parse();
                let elapsed = start_time.elapsed();
                
                match result {
                    Ok(json) => {
                        assert!(json.is_valid());
                        file_elements = json.size();
                        file_time += elapsed;
                    }
                    Err(error) => {
                        println!("❌ JSON parsing failed: {}", error);
                        return Err(error.into());
                    }
                }
            }
            
            let avg_time = file_time / iterations as u32;
            let bytes_per_second = if avg_time.as_secs_f64() > 0.0 {
                file_size as f64 / avg_time.as_secs_f64()
            } else {
                0.0
            };
            
            println!("  Elements: {}, Avg Time: {:?}, Bytes/sec: {:.0}", 
                     file_elements, avg_time, bytes_per_second);
            
            total_elements += file_elements as u64;
            total_bytes += file_size as u64;
            total_time += avg_time;
        } else {
            println!("Warning: Test file {} not found, skipping...", file_path);
        }
    }
    
    let overall_bytes_per_sec = if total_time.as_secs_f64() > 0.0 {
        total_bytes as f64 / total_time.as_secs_f64()
    } else {
        0.0
    };
    
    println!("Rust JSON Parser Overall Performance:");
    println!("  Total elements: {}", total_elements);
    println!("  Total bytes: {}", total_bytes);
    println!("  Total time: {:?}", total_time);
    println!("  Average bytes/second: {:.0}", overall_bytes_per_sec);
    println!("  Average MB/sec: {:.2}", overall_bytes_per_sec / (1024.0 * 1024.0));
    
    Ok(())
}

fn generate_deeply_nested_json(depth: usize) -> String {
    let mut json = String::new();
    
    for _ in 0..depth {
        json.push_str("{\"nested\":");
    }
    
    json.push_str("\"value\"");
    
    for _ in 0..depth {
        json.push('}');
    }
    
    json
}

fn generate_wide_json(count: usize) -> String {
    let mut json = String::from("{");
    
    for i in 0..count {
        if i > 0 {
            json.push(',');
        }
        json.push_str(&format!("\"key{}\": {}", i, i));
    }
    
    json.push('}');
    json
}

pub fn benchmark_json_parser_stress_test() -> Result<(), Box<dyn std::error::Error>> {
    println!("Running Rust JSON parser stress tests...");
    
    // Test deeply nested structures
    let deeply_nested = generate_deeply_nested_json(1000);
    let static_nested: &'static str = Box::leak(deeply_nested.into_boxed_str());
    
    let start = Instant::now();
    let mut parser = JsonParser::new(static_nested);
    let result = parser.parse()?;
    let elapsed = start.elapsed();
    
    assert!(result.is_valid());
    println!("  Deeply nested (1000 levels): {:?}", elapsed);
    
    // Test wide structures
    let wide_structure = generate_wide_json(10000);
    let static_wide: &'static str = Box::leak(wide_structure.into_boxed_str());
    
    let start = Instant::now();
    let mut parser = JsonParser::new(static_wide);
    let result = parser.parse()?;
    let elapsed = start.elapsed();
    
    assert!(result.is_valid());
    println!("  Wide structure (10000 keys): {:?}", elapsed);
    
    Ok(())
}

fn main() {
    println!("Running Rust JSON Parser Benchmarks...");
    
    if let Err(e) = benchmark_json_parser_real_world() {
        eprintln!("Error running real-world JSON parser benchmark: {}", e);
        std::process::exit(1);
    }
    
    if let Err(e) = benchmark_json_parser_stress_test() {
        eprintln!("Error running JSON parser stress test: {}", e);
        std::process::exit(1);
    }
    
    println!("Rust JSON parser benchmarks completed successfully!");
}